File: AlgorithmsPackPythonSolutions/300. Longest Increasing Subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        MAX = 2500 + 1
        memory = [-1] * MAX
        
        def LIS(index):
            if index >= len(nums):
                return 0
            if memory[index] != -1:
                return memory[index]
            ret = 0
            for next_index in range(index, len(nums)):
                if nums[index] < nums[next_index]:
                    ret = max(ret, LIS(next_index))
            ret += 1
            memory[index] = ret
            return ret
        
        nums = [float('-inf')] + nums
        return LIS(0) - 1

File: AlgorithmsPackCPP/300. Longest Increasing Subsequence.cpp
const int MAX = 2500+1;
int memory[MAX];
vector<int>nums;

class Solution {
public:
    int LIS(int index){
        if(index>=nums.size())
            return 0;
        auto&ret = memory[index];
        if(ret!=-1)
            return ret;
        ret = 0;
        for(int next = index; next < nums.size(); next++){
            if(nums[index]<nums[next])
                ret = max(ret,LIS(next));
        }
        ret+=1;
        return ret;
    }

    int lengthOfLIS(vector<int>& nums_) {
        nums = nums_;
        nums.insert(nums.begin(),INT_MIN);
        memset(memory,-1,sizeof(memory));
        return LIS(0)-1;
    }
};